receiver/kafkareceiver/kafka_receiver.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2020 The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafkareceiver
+
+import (
+	"context"
+
+	"github.com/Shopify/sarama"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/obsreport"
+)
+
+const (
+	transport = "kafka"
+)
+
+// kafkaConsumer uses sarama to consume and handle messages from kafka.
+type kafkaConsumer struct {
+	name          string
+	consumerGroup sarama.ConsumerGroup
+	nextConsumer  consumer.TraceConsumer
+	topics        []string
+	cancel        context.CancelFunc
+	unmarshaller  unmarshaller
+
+	logger *zap.Logger
+}
+
+var _ component.Receiver = (*kafkaConsumer)(nil)
+
+func newReceiver(config Config, params component.ReceiverCreateParams, nextConsumer consumer.TraceConsumer) (*kafkaConsumer, error) {
+	c := sarama.NewConfig()
+	c.ClientID = config.ClientID
+	c.Metadata.Full = config.Metadata.Full
+	c.Metadata.Retry.Max = config.Metadata.Retry.Max
+	c.Metadata.Retry.Backoff = config.Metadata.Retry.Backoff
+	if config.ProtocolVersion != "" {
+		version, err := sarama.ParseKafkaVersion(config.ProtocolVersion)
+		if err != nil {
+			return nil, err
+		}
+		c.Version = version
+	}
+	client, err := sarama.NewConsumerGroup(config.Brokers, config.GroupID, c)
+	if err != nil {
+		return nil, err
+	}
+	return &kafkaConsumer{
+		name:          config.Name(),
+		consumerGroup: client,
+		topics:        []string{config.Topic},
+		nextConsumer:  nextConsumer,
+		unmarshaller:  &protoUnmarshaller{},
+		logger:        params.Logger,
+	}, nil
+}
+
+func (c *kafkaConsumer) Start(context.Context, component.Host) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	c.cancel = cancel
```

Comment:
Confused about this cancel, is this just to cancel the "consumerGroup.Consume"? Is that not cancelled by the `Close`?

Reply:
It's used to interrupt the infinite loop that calls `c.consumerGroup.Consume`